File: Data Structures Intro/Disjoint Sets/disjoint-set-task-net/DisjointSet/DisjointSets.cs
using DisjointSet.Interfaces;

namespace DisjointSet.DisjointSet;

/// <summary>
/// Implementation of disjoint sets data structure with path compression and union by rank optimizations.
/// </summary>
/// <typeparam name="T">Type of elements in the sets</typeparam>
public class DisjointSets<T> : IDisjointSets<T> where T : notnull
{
    // parent хранит "родителя" для каждого элемента
    private readonly Dictionary<T, T> _parent = new();
    // rank хранит "высоту дерева" (или размер множества)
    private readonly Dictionary<T, int> _rank = new();

    /// <summary>
    /// Creates a new set that is associated with a given key.
    /// </summary>
    /// <param name="key">The key to create a set for</param>
    /// <exception cref="ArgumentNullException">Thrown when key is null</exception>
    public void MakeSet(T key)
    {
        if (key is null)
            throw new ArgumentNullException(nameof(key));

        if (_parent.ContainsKey(key))
            return; // уже существует — игнорируем

        _parent[key] = key;
        _rank[key] = 0; // изначально "ранг" = 0
    }

    /// <summary>
    /// Returns a unique set identifier (key) of a given's key set.
    /// Uses path compression.
    /// </summary>
    /// <param name="key">The key to find the set for</param>
    /// <returns>The unique set identifier</returns>
    /// <exception cref="KeyNotFoundException">Thrown when key was not added by MakeSet</exception>
    public T FindSet(T key)
    {
        if (!_parent.ContainsKey(key))
            throw new KeyNotFoundException($"Key '{key}' was not found.");

        if (!EqualityComparer<T>.Default.Equals(_parent[key], key))
            _parent[key] = FindSet(_parent[key]); // рекурсивная компрессия пути

        return _parent[key];
    }

    /// <summary>
    /// Joins two given sets into a new one using union by rank heuristic.
    /// If ranks are equal, prefer second set as the root.
    /// </summary>
    public void UnionSets(T firstKey, T secondKey)
    {
        T root1 = FindSet(firstKey);
        T root2 = FindSet(secondKey);

        if (EqualityComparer<T>.Default.Equals(root1, root2))
            return; // уже в одном множестве

        int rank1 = _rank[root1];
        int rank2 = _rank[root2];

        if (rank1 < rank2)
        {
            _parent[root1] = root2;
        }
        else if (rank1 > rank2)
        {
            _parent[root2] = root1;
        }
        else
        {
            // ранги равны — выбираем второй как root (согласно комментарию в задании)
            _parent[root1] = root2;
            _rank[root2]++; // увеличиваем ранг root2
        }
    }
}

File: Data Structures Intro/Disjoint Sets/Java/disjoint-set-task-java/src/main/java/com/epam/bsp/disjoint/set/DisjointSets.java
package com.epam.bsp.disjoint.set;

import java.util.HashMap;
import java.util.Map;

/**
 * Class for supporting disjoint sets.
 * NOTE: the expected implementation should contain:
 * "path compression" heuristic
 * "union by rank" heuristic (based on sets size)
 *
 * @param <K> type of key
 */
public class DisjointSets<K> {

    private Map<K, K> parent; // parent map
    private Map<K, Integer> size; // size (rank) map

    public DisjointSets() {
        parent = new HashMap<>();
        size = new HashMap<>();
    }

    /**
     * Creates a new set that is associated with a given key.
     * @param key the key
     */
    public void makeSet(K key) {
        parent.put(key, key);
        size.put(key, 1);
    }

    /**
     * Returns a unique set identifier (key) of a given's key set.
     * NOTE: "path compression" heuristic should be used.
     *
     * @param key the key
     * @return the unique set identifier
     */
    public K findSet(K key) {
        K p = parent.get(key);
        if (p == null) return null; // key not found
        if (!p.equals(key)) {
            parent.put(key, findSet(p)); // path compression
        }
        return parent.get(key);
    }

    /**
     * Joins two given sets into a new one.
     * NOTE: "union by rank" heuristic should be used (based on sets size).
     * NOTE: if the sets that correspond to the given keys are of the same rank,
     * it is preferable to use the second set when deciding what set is to be used as a new 'root'.
     *
     * @param firstKey the key of a first set
     * @param secondKey the key of a second set
     */
    public void unionSets(K firstKey, K secondKey) {
        K root1 = findSet(firstKey);
        K root2 = findSet(secondKey);

        if (root1 == null || root2 == null) return;
        if (root1.equals(root2)) return; // already in same set

        int size1 = size.get(root1);
        int size2 = size.get(root2);

        // union by size (rank)
        if (size1 > size2) {
            parent.put(root2, root1);
            size.put(root1, size1 + size2);
        } else {
            parent.put(root1, root2);
            size.put(root2, size1 + size2);
        }
    }
}
